tests/test_timestamp.py

```diff
@@ -0,0 +1,138 @@
+
+
+class _P(record('timestamps', 'expected_microseconds')):
+    def __str__(self):
+        return self.desc
+
+
+MISSING_MICROSECOND = _P(
+    timestamps=[Timestamp(
+            2011, 1, 1,
+            0, 0, 0, None,
+            precision=TimestampPrecision.SECOND, fractional_precision=6, fractional_seconds=0
+            ),
+            Timestamp(
+            2011, 1, 1,
+            0, 0, 0, None,
+            precision=TimestampPrecision.SECOND, fractional_precision=6, fractional_seconds=Decimal('0.123456')
+            ),
+            Timestamp(
+            2011, 1, 1,
+            0, 0, 0, None,
+            precision=TimestampPrecision.SECOND, fractional_precision=6, fractional_seconds=Decimal('0.123456789')
+            ),
+            Timestamp(
+            2011, 1, 1,
+            0, 0, 0, None,
+            precision=TimestampPrecision.SECOND, fractional_precision=6, fractional_seconds=Decimal('0.000123')
+            ),
+            Timestamp(
+            2011, 1, 1,
+            0, 0, 0, None,
+            precision=TimestampPrecision.SECOND, fractional_precision=6, fractional_seconds=Decimal('0.123000')
+            )],
+    expected_microseconds=[0, 123456, 123456, 123, 123000]
+)
+
+
+@parametrize(
+    MISSING_MICROSECOND
+)
+def test_missing_microsecond(item):
+    for i in range(len(item.timestamps)):
+        assert item.timestamps[i].microsecond == item.expected_microseconds[i]
+
+
+class _P(record('timestamps', 'expected_fractional_precision')):
+    def __str__(self):
+        return self.desc
+
+
+MISSING_FRACTIONAL_PRECISION = _P(
+    timestamps=[Timestamp(
+            2011, 1, 1,
+            0, 0, 0, 0,
+            precision=TimestampPrecision.SECOND, fractional_precision=6, fractional_seconds=None
+            ),
+            Timestamp(
+            2011, 1, 1,
+            0, 0, 0, 123456,
+            precision=TimestampPrecision.SECOND, fractional_precision=6, fractional_seconds=None
+            ),
+            Timestamp(
+            2011, 1, 1,
+            0, 0, 0, 123,
+            precision=TimestampPrecision.SECOND, fractional_precision=6, fractional_seconds=None
+            ),
+            Timestamp(
+            2011, 1, 1,
+            0, 0, 0, 123000,
+            precision=TimestampPrecision.SECOND, fractional_precision=6, fractional_seconds=None
+            )],
+    expected_fractional_precision=[0, Decimal('0.123456'), Decimal('0.000123'), Decimal('0.123000')]
+)
+
+
+@parametrize(
+    MISSING_FRACTIONAL_PRECISION
+)
+def test_missing_fractional_seconds(item):
+    for i in range(len(item.timestamps)):
+        assert item.timestamps[i].fractional_seconds == item.expected_fractional_precision[i]
+
+
+def test_nonequivalent_microsecond_and_fractional_seconds():
+    with pytest.raises(ValueError):
+        Timestamp(
+            2011, 1, 1,
+            0, 0, 0, 123456,
+            precision=TimestampPrecision.SECOND, fractional_precision=6, fractional_seconds=0
+        )
```

Comment:
Minor: I would collapse these into one set of data driven tests, but it isn't a big deal to have them each spelled out.

Reply:
Done, collapsed into one.

Reply:
No, perhaps I was not clear, I didn't mean collapse your assertions into one test. I meant that you should have a case per each driven as a **parameterized** test.  The problem with your refactoring is that if all of these fail it only looks like one failure.  If you don't want to make it parameterized, keep each test separate.

Reply:
Done. Each test kept separate.